Support for weights for MAP and Within Algorithms

s3alfisc opened the pull request:
This PR adds support for weighted demeaning for the MAP and Within Algorithm. 

It 
- adds a `weights` argument to the `residualize` method
- `interface.Group.mean` has been updated to compute weighted means via `np.bincount`
- all `_residualize_matrix()` methods obtain a new argument `weights`. within the methods, I check if weighted demean is supported. 

To Do's (I will do this over the weekend): 

- [ ] add support for the `Dummy` method
- [ ] add testing

What I will not do: 
- add support for the other algorithms, as I do not know / understand them well enough. 

Here's a short code example: 

```py
%reload_ext autoreload
%autoreload 2

import pyhdfe 
import numpy as np

N = 1000
X = np.random.normal(0, 1, 2*N).reshape((N, 2))
g = np.random.choice(["a", "b", "c", "d", "e"], N).reshape((N, 1))
weights = np.random.uniform(0, 1, N).reshape((N, 1))

within = pyhdfe.create(g, residualize_method="within").residualize(X, weights = weights)
map = pyhdfe.create(g, residualize_method="map").residualize(X, weights = weights)

map_weights = pyhdfe.create(g, residualize_method="map").residualize(X, weights = np.ones(N).reshape((N, 1)))
map_noweights = pyhdfe.create(g, residualize_method="map").residualize(X)

print(np.allclose(within, map))
# True
print(np.allclose(map_weights, map_noweights))
# True
```

s3alfisc committed:
add initial support for weights

jeffgortmaker commented:
Great, thanks. I'll go over the code you do have now.

jeffgortmaker commented on pyhdfe/algorithms.py:
```diff
@@ -134,7 +134,7 @@ def _compute_degrees(self, cluster_ids: Optional[Array], degrees_method: Optiona
         return degrees, singletons
 
-    def residualize(self, matrix: Array) -> Array:
+    def residualize(self, matrix: Array, weights: Optional[Array] = None) -> Array:
```
You'll need docs too!

jeffgortmaker commented on pyhdfe/algorithms.py:
```diff
@@ -169,10 +169,23 @@ def residualize(self, matrix: Array) -> Array:
-        return self._residualize_matrix(matrix)
+
+        if weights is not None:
+            #print(type(self))
```
Probably easiest to raise a separate exception within algorithm classes / option configurations that are not yet supported, saying explicitly that <algorithm / option> name is not supported with weights.

s3alfisc replied:
I now raise `NotImplementedErrors` whenever algo's do not support weights.

jeffgortmaker commented on pyhdfe/algorithms.py:
```diff
@@ -169,10 +169,23 @@ def residualize(self, matrix: Array) -> Array:
 
     @abc.abstractmethod
-    def _residualize_matrix(self, matrix: Array) -> Array:
+    def _residualize_matrix(self, matrix: Array, weights: Union[Array, None]) -> Array:
```
Use `weights: Optional[Array] = None` instead of `Union`. Setting a default `None` means this can be used for non-weighted things too without passing a dummy `None`.

jeffgortmaker commented on pyhdfe/algorithms.py:
```diff
@@ -188,21 +201,22 @@ def __init__(
-    def _residualize_matrix(self, matrix: Array) -> Array:
+    def _residualize_matrix(self, matrix: Array, weights: Union[Array, None]) -> Array:
         """Compute residuals from regressions of each matrix column on the dummy variables."""
+        if weights is not None:
```
Great, yeah, something like this for each non-supported algorithm. Adding a test that this works too, would be useful. I.e. we should get an exception for each non-supported algorithm / option config.

jeffgortmaker replied:
Also, the right exception class is `NotImplementedError`.

s3alfisc replied:
Fixed all cases where weights are not supported to draw a `NotImplementedError`.

jeffgortmaker commented on pyhdfe/utilities.py:
```diff
@@ -89,9 +89,16 @@ def sum(self, matrix: Array) -> Array:
-        return self.sum(matrix) / self.counts[:, None]
+        if weights is None:
+            return self.sum(matrix) / self.counts[:, None]
+        else:
```
Don't need an else statement after a `return`

s3alfisc replied:
Done.

jeffgortmaker commented on pyhdfe/utilities.py:
```diff
@@ -89,9 +89,16 @@ def sum(self, matrix: Array) -> Array:
+        if weights is None:
+            return self.sum(matrix) / self.counts[:, None]
+        else:
+            if not hasattr(self, "wcounts"):
```
A few coding style notes here that probably won't matter after some other suggested changes.

1. Setting attributes like this in an ad-hoc manner isn't very Pythonic. Better is to set attributes by default `None` in `__init__` with documentation about what they do, when configured.
2. Generally speaking I've tried to use longer variable names throughout the codebase, unless using math symbol names. No need to abbreviate to `wcounts`.

s3alfisc replied:
I think this should no longer be relevant, so I am setting it as resolved.

jeffgortmaker commented on pyhdfe/utilities.py:
```diff
@@ -89,9 +89,16 @@ def sum(self, matrix: Array) -> Array:
+                # note: for np.bincount to work, self.codes need to be a sequence of
+                # integers from 0 to max(self.codes), without 'holes'
+                self.wcounts = np.bincount(self.codes, weights.flatten())
+            return self.sum( matrix * weights) / self.wcounts[:, None]
```
I'm a bit confused about this implementation. Does dividing by `self.sum(weights)` not work?

More generally, mind thinking a bit more about what `weights` should actually be? If for most use-cases they'll already be normalized to sum to one within groups, then we can make this expectation clear in the documentation and not divide by `self.sum(weights)` here, nor have to worry about caching them. Since this package is supposed to be a utilities package used by other packages with nicer interfaces, I'd be okay with this.

If we really want to allow for weights to not already sum to one within groups, one possibility is to normalize them at the beginning of the top-level `residualize` method. Although we'd probably want to have a `normalize_weights: bool = True` argument so that people can set this to `False` if the weights they're passing are already normalized and they want to avoid this extra computation time. But again, I think I prefer 1? Let me know what you think.

One possible issue is if pre-normalizing weights might lead to some numerical instability issues. Actually my intuition is that pre-normalizing them (or requiring them to be pre-normalized) will probably help with numerical instability, but it'd also probably be good to check what other implementations are doing (e.g. `reghdfe`'s).

s3alfisc replied:
Hi Jeff, sorry for taking a while, work + some R package duties kept my from continuing with this PR. 

>  I'm a bit confused about this implementation. Does dividing by self.sum(weights) not work?

Yes, I don't think that it works. Here are some examples: 

```py
import numpy as np
import pyhdfe

N = 1000
weights = np.random.choice([0, 1, 2, 3, 4], N,  True).reshape((N, 1))
f = np.random.choice([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], N, True).reshape((N, 1))
matrix = np.random.normal(0, 1, N*3).reshape((N, 3 ))

algorithm = pyhdfe.create(f)
groups = algorithm._groups_list[0]

# for each of the covariates, sum by one of the two fixef groups -> n_fe x k matrix
groups.sum(matrix)
array([[ 18.21773928,   2.01516345,  -2.91394769],
       [-16.63034115,   6.1069725 ,  20.84053581],
       [  7.59278491,  13.05964635,   2.01013214],
       [ -1.91232025,  13.49004955,  15.72778202],
       [ -5.76620718,  19.92604936,  -9.21337829],
       [ -1.03836189,   4.02464604,  -4.81444732],
       [  3.10541319,   8.41333695,  -3.35792013],
       [ -5.85394972,   4.33906534,   3.08187546],
       [ 15.58773835, -15.23807211,  -4.28672576],
       [  9.4700185 ,   0.34346523,   5.13255758]])

# now compute means by dividing via count
groups.sum(matrix) / groups.counts[:, None]
array([[ 0.18976812,  0.02099129, -0.03035362],
       [-0.15990713,  0.05872089,  0.20038977],
       [ 0.07827613,  0.13463553,  0.02072301],
       [-0.01931637,  0.13626313,  0.15886649],
       [-0.05148399,  0.17791115, -0.08226231],
       [-0.01070476,  0.0414912 , -0.04963348],
       [ 0.02902255,  0.07862932, -0.03138243],
       [-0.06362989,  0.04716375,  0.03349865],
       [ 0.15282096, -0.14939286, -0.04202672],
       [ 0.10074488,  0.00365389,  0.05460168]])
```

Equivalent results can be produced by using `np.bincount`: 

```py
groups.sum(matrix) / np.bincount(groups.codes)[:,None]

array([[ 0.18976812,  0.02099129, -0.03035362],
       [-0.15990713,  0.05872089,  0.20038977],
       [ 0.07827613,  0.13463553,  0.02072301],
       [-0.01931637,  0.13626313,  0.15886649],
       [-0.05148399,  0.17791115, -0.08226231],
       [-0.01070476,  0.0414912 , -0.04963348],
       [ 0.02902255,  0.07862932, -0.03138243],
       [-0.06362989,  0.04716375,  0.03349865],
       [ 0.15282096, -0.14939286, -0.04202672],
       [ 0.10074488,  0.00365389,  0.05460168]])
```

For weighted means, we can use bincounts weights option: 

```py
groups.sum(matrix * weights) / np.bincount(f.flatten(), weights.flatten())[:, None]
array([[ 0.22557294,  0.01113643,  0.02486263],
       [-0.07056653,  0.05479159,  0.32698637],
       [ 0.07438351,  0.02514269,  0.04918748],
       [-0.01848661,  0.20785587,  0.07793799],
       [-0.04196025,  0.19014639, -0.02457532],
       [ 0.02809851,  0.0911911 , -0.14453099],
       [ 0.07275208,  0.05682283, -0.04452464],
       [-0.16254244,  0.05435781,  0.0286162 ],
       [ 0.18282444, -0.10016904,  0.11724293],
       [ 0.09508381, -0.03571426, -0.02568161]])
```
which is different from 

```py
groups.sum(matrix * weights) / np.sum(weights)

array([[ 0.02280692,  0.00112597,  0.00251378],
       [-0.00702826,  0.00545711,  0.03256705],
       [ 0.00763292,  0.00258003,  0.00504741],
       [-0.00182262,  0.02049283,  0.00768403],
       [-0.00434799,  0.0197033 , -0.00254654],
       [ 0.00277028,  0.00899067, -0.01424953],
       [ 0.00742891,  0.00580233, -0.00454653],
       [-0.01618883,  0.00541391,  0.0028501 ],
       [ 0.01866869, -0.01022853,  0.01197199],
       [ 0.00875268, -0.00328758, -0.00236405]])

```
as `np.sum()` does not "count" by fixed effects group. 

Regarding the weight scale: In the R world, it is standard to allow for weights that do not sum to one, so I guess that is why I did not put any constraint on this. I will add a `normalize_weights` argument and will check what `fixest` and `reghdfe` do to inform us on which default to choose.

Best, Alex

jeffgortmaker replied:
My suggestion was to do `groups.sum(matrix * weights) / groups.sum(weights)`. That gives you the desired result and should be faster / doesn't require new code.

And sounds good regarding checking what `fixest` and `reghdfe` do here!

s3alfisc replied:
Yes, of course groups.sum(weights) works, I'm in mild disbelief I did not figure this out myself 😅

s3alfisc replied:
I have implemented the suggested approach.

s3alfisc commented:
> Great, thanks. I'll go over the code you do have now.

Thanks for your feedback, Jeff. I will work on this tomorrow =) Best, Alex

s3alfisc committed:
add docs

s3alfisc committed:
add docs for residualize

s3alfisc committed:
union -> optional

s3alfisc committed:
drop superfluous else

s3alfisc committed:
rework - weights as arg to create

s3alfisc committed:
support for WLS for Dummy Class

s3alfisc committed:
support for WLS for Dummy Class

s3alfisc committed:
minor documentation improvement

s3alfisc committed:
first attempt at tests for weighed means

s3alfisc committed:
add CI

s3alfisc committed:
fix some test bugs

s3alfisc committed:
add workflows folder

s3alfisc committed:
install pytest

s3alfisc committed:
no weights with cg and gk acceleration

s3alfisc commented:
Hi @jeffgortmaker, I hope this finds you well!

I finally got around to give this PR another push! I rewrote the implementation - users now have to provide the weights when initiating the algorithm, and not longer have to provide it to the `residualize` method. The key argument for doing so was implementation convenience - this way requires fewer lines of code, as I can check for `weights` support in `interface()` (vs the individual algorithm implementations). Let me know if you don't like this choice, I am more than happy to revert it. 

One argument against it is that for glm's fitted via iterative weighted least squares, the algorithm has to be initiated multiple times (as weights are updated, see here for [my implementation of pplmhdfe](https://github.com/s3alfisc/pyfixest/blob/c2f25b1e3ef4cda5990bfb6b9de88b3b7fe7a7fd/pyfixest/fepois.py#L114)). Though I believe this adds mostly insignificant overhead, and I can also simply opt to overwrite the weights attribute of a given instance of an `algorithm`. 

The algorithm class now gets a `weights` attribute, which is either provided by the user or a vector of ones. I have updated the `Dummy` class to support WLS (which is of course required for testing). 

Currently, the implementation does not allow for the `cg` and `gk` accelerations with the MAP algo, as I ran into some convergence issues when testing (approx 15 tests failed because the iteration limit was reached). 

I have added continuous integration to run the tests, as I thought this was the easiest way to share the test results with you (I am more than happy to take it out again in case you do not want it, though I personally find it very useful). Currently, one out of 14K tests fails as it does not meet the tolerance level (atol and rtol are at 3e-12 > 1e-12). Hence lowering the tolerance slightly will fix this. 

Regarding the question of standardizing weights, I have not yet gotten a clear answer. In their `pplmhdfe` paper, Correia et al do not mention standardization of their iterative weighted least squares weights, though that of course does not mean they don't do it in their implementation. 

Best, Alex

s3alfisc committed:
minor cleanups

jeffgortmaker commented:
Fantastic, thanks! I’ll be able to get to this early this week, I think.

jeffgortmaker commented:
Okay, went over the changes.

1. Your one example where one would have to initialize multiple algorithms is enough to convince me that it is definitely better to have `weights` be an argument to `Algorithm.residualize`, not `create`. A primary goal of this package is to minimize overhead, so this is sort of first-order. A simple way to do this without much code is to have a new `Algorithm._supports_weights` attribute, which is by default `False`, but which in specific algorithms' `__init__` you can set to `True`, if they support weights. Then you'd have error checking for this in the abstract `Algorithm.residualize` method.
2. Building on this, I did some research, and it seems like generally speaking, standardizing weights before computing a weighted average spreads out the division operator among many elements, resulting in improved numerical stability. This along with the fact that for many users, weights may *already* be standardized (so re-dividing by their sum would create unnecessary overhead) suggests that the best approach is to actually have the new `Algorithm.residualize` argument be `standardized_weights`. And we'd be explicit in the documentation that we're assuming these already sum to 1 (we won't check this because this would add overhead). So we then don't have to divide by their sum anywhere, or even keep track of this sum.
3. Can you remove the check that `weights <= 0`? This creates additional overhead, and we can assume weights have already been cleaned. Remember this isn't supposed to be a user-friendly package, just a backend for other packages ;). Also there might be some cases where we actually want negative weights -- not sure!
4. With minimal overhead in mind, please special-case having weights instead of having them default to a vector of ones. A vector of ones takes up a good amount of memory for all the cases when we don't have weights at all. So weights should be `None` by default, taking up essentially no memory, and only if they're non-`None`, you should do weighted stuff (e.g. in `Dummy`, multiplying by their square root).
5. Not allowing for weights with different algorithms is okay, just be explicit about what's supported in the `weights` documentation. I am curious why these are failing, but probably something to figure out later!
6. No need to commit your CI file -- I can check out your PR locally and run tests myself! Feel free to add `.github` to the `.gitignore` if it helps you work on this, though.

s3alfisc committed:
3. no weights should be positive check

s3alfisc committed:
6. delete CI

s3alfisc commented:
Great, thanks for the feedback! I have already started to work on it, let's see how far I will make it tonight =)

s3alfisc committed:
1. move weights arg from create to residualize - basically jump back to the initial implementation + add self._support_weights attr

s3alfisc commented:
I have now moved the weights argument back from `create` to `residualize` + implemented checks via the `self._support_weights` attribute. Basically, we're back to my initial PR 😄  [864ca6b](https://github.com/jeffgortmaker/pyhdfe/pull/4/commits/864ca6bcf87d29c06d3925a8a4dc6a8114cf1c1a) Funnily, when reading your message, I recalled that one of the main reasons for moving `weights` from `residualize` to `create` was that I did not see a simple solution for `NotImplementedError` error handling, so thanks for the suggestion! All tests pass as before, with the one exception mentioned above. All in all, I think that having `weights` as an argument to `residualize` makes for a nicer API. 

So 1, 3, 4, and 6 should be done, and tomorrow I'll add 2 and 5, check if I forgot anything that you mentioned, go through the documentation and clean up the code wherever necessary.

jeffgortmaker commented:
Fantastic! For the test that's not passing, you can absolutely add a case that checks if weights are there, and if so, loosens the tolerance a bit. (For code cleanup, running `flake8` and `mypy .` in the top-level repo will make sure style and typing match with the rest of the package.)

s3alfisc committed:
2. add standardize_weights arg, which defaults to False. update tests. add docs on weights

s3alfisc committed:
delete debug.py

s3alfisc committed:
run flake8 and update

s3alfisc committed:
fix some type errors

s3alfisc committed:
add flake8 fix

s3alfisc commented:
I have added a standardize weights error and fixed all complaints of `flake8`. 

Regarding `mypy`, I get a bunch of errors that I think (?) are unrelated to my PR, or if they are related, errors that you seem to have been fine with in previous versions of `pyhdfe` (under slightly different circumstances). 

I have not lowered the tolerance for tests with weights - as it turns out, the error occurs for an unweighted problem: 

```py
FAILED tests\test_hdfe.py::test_algorithms[covariates: 13-scales: (2, 2, 2, 2)-levels: (4, 4, 4, 4)-singletons: 10%-weights: False-standardize weights-LSMR-residual-singletons retained] - AssertionError:
FAILED tests\test_hdfe.py::test_algorithms[covariates: 13-scales: (2, 2, 2, 2)-levels: (4, 4, 4, 4)-singletons: 10%-weights: False-raw weights-LSMR-residual-singletons retained] - AssertionError:
```

jeffgortmaker commented:
Great, I think I’ll be able to get to this by Tuesday.

jeffgortmaker commented:
Okay, this looks pretty good! I'm going to merge the PR and deal with any remaining small things on my own.

One non-small thing that I'm going to remove standardization. I just realized from reading your code that I had been misunderstanding the whole issue with standardizing the weights. Sorry about that. Since we're taking weighted averages within different groups for different IDs, making the weights sum to 1 across the whole dataset doesn't really do the trick that I was hoping for (being able to just sum matrix * weights). Instead, we'd have to pre-compute a different vector of weights for each column of IDs. I'm going to leave that for some future PR if someone really wants to speed up this part of the package. Let me know if that makes sense, or if I'm missing something else here.

Thanks for all your work on this! It's a great contribution to the package.

s3alfisc commented:
Amazing! Thanks for guiding me through the PR - I've learned quite a lot of Python tricks while working on this! =)
